src/transport/raw/TCP.cpp

```diff
@@ -381,7 +377,7 @@ CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const Pe
             //   - there is no reason to believe that an error would not occur again on the
             //     same parameters (errors are likely not transient)
             //   - this guarantees data is received and progress is made.
-            err = ProcessSingleMessageFromBufferHead(peerAddress, buffer, messageSize);
+            err = ProcessSingleMessageFromBufferHead(peerAddress, buffer.Retain(), messageSize);
```

Comment:
Ah, ok, we're doing the extra `Retain` here to get the right semantics.  I still think it would be clearer if callee took a reference instead.